components/style/rule_tree/map.rs

```diff
@@ -0,0 +1,164 @@
+        match &self.inner {
+            MapInner::One(one) if *key == key_from_value(one) => Some(one),
+            MapInner::Map(map) => map.get(key),
+            MapInner::Empty | MapInner::One(_) => None,
+        }
+    }
+
+    pub(super) fn get_or_insert_with(
+        &mut self,
+        key: K,
+        key_from_value: impl FnOnce(&V) -> K,
+        new_value: impl FnOnce() -> V,
+    ) -> &mut V {
+        match self.inner {
+            MapInner::Empty => {
+                self.inner = MapInner::One(new_value());
+                match &mut self.inner {
+                    MapInner::One(one) => one,
+                    _ => unreachable!(),
+                }
+            },
+            MapInner::One(_) => {
+                let one = match mem::replace(&mut self.inner, MapInner::Empty) {
+                    MapInner::One(one) => one,
+                    _ => unreachable!(),
+                };
+                // If this panics, the child `one` will be lost.
+                let one_key = key_from_value(&one);
+                // Same for the equality test.
+                if key == one_key {
+                    self.inner = MapInner::One(one);
+                    match &mut self.inner {
+                        MapInner::One(one) => return one,
+                        _ => unreachable!(),
+                    }
+                }
+                self.inner = MapInner::Map(Box::new(FxHashMap::with_capacity_and_hasher(
+                    2,
+                    Default::default(),
+                )));
+                let map = match &mut self.inner {
+                    MapInner::Map(map) => map,
+                    _ => unreachable!(),
+                };
+                map.insert(one_key, one);
+                // But it doesn't matter if f panics, by this point
+                // the map is as before but represented as a map instead
+                // of a single value.
+                map.entry(key).or_insert_with(new_value)
+            },
+            MapInner::Map(ref mut map) => map.entry(key).or_insert_with(new_value),
+        }
+    }
+
+    pub(super) fn remove(&mut self, key: &K, key_from_value: impl FnOnce(&V) -> K) -> Option<V> {
+        match &mut self.inner {
+            MapInner::One(one) if *key == key_from_value(one) => {
+                match mem::replace(&mut self.inner, MapInner::Empty) {
+                    MapInner::One(one) => Some(one),
+                    _ => unreachable!(),
+                }
+            },
+            MapInner::Map(map) => map.remove(key),
+            MapInner::Empty | MapInner::One(_) => None,
+        }
+    }
+}
+
+impl<K, V> MallocShallowSizeOf for Map<K, V>
+where
+    K: Eq + Hash,
+{
+    fn shallow_size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+        match &self.inner {
+            MapInner::Map(m) => m.shallow_size_of(ops) + (**m).shallow_size_of(ops),
```

Comment:
Maybe keep the `// We want to account for both the box and the hashmap.` as it's not quite obvious why the double deref is needed.

Reply:
Also, fwiw, I'd rather do `match self.inner` and `Map(ref m)` than having to mint it in my head, but your call :)